### ISSUE TITLE
Poor shuffle implementation for DocumentData::Shuffle?

### ISSUE THREAD

**Issue by jeffhube** [tesseract-ocr/tesseract#3316]
### Environment

* **Tesseract Version**: ff830775f959ac8e13eec908f02c16459a2ffb6d
* **Commit Number**: ff830775f959ac8e13eec908f02c16459a2ffb6d
* **Platform**: Linux ubuntuvm 5.8.0-44-generic #50-Ubuntu SMP Tue Feb 9 06:29:41 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux

### Current Behavior:

I noticed that DocumentData::Shuffle() here https://github.com/tesseract-ocr/tesseract/blob/ff830775f959ac8e13eec908f02c16459a2ffb6d/src/ccstruct/imagedata.cpp#L522-L535 is using a naive shuffle algorithm. Testing indicates that approximately 13% of items are in their starting position after shuffling with this algorithm. I don't know if this has any noticeable impact on the behavior of tesseract but I figured I might as well mention it.

### Expected Behavior:

Shuffling is performed using the Fisher–Yates shuffle algorithm.

### Suggested Fix:

"

I would be willing to make a PR later if this is worth fixing.

**Comment by stweil** [comment]
The `tesstrain` repository which is used for training with real line images uses a [different code for shuffling](https://github.com/tesseract-ocr/tesstrain/blob/master/shuffle.py).

The code here is used for the training with artificial line images. I think it might be worth to add a better shuffle algorithm. For compatibility reasons the old one should be supported as an alternative for some time.

**Comment by jeffhube** [comment]
So would this need a flag passed to `tesstrain.sh` which is then passed on to `tesseract` to enable the new behavior?

**Comment by stweil** [comment]
Yes, it would require a new parameter to select the shuffle algorithm. The new algorithm could be the default.

Maybe it is possible to use `std::random_shuffle` or `std::shuffle` for the implementation.

**Comment by egorpugin** [comment]
Is it possible to use something from C++ stdlib for shuffling for tess code (original post)?
